File: demos/two_process/supplier_agent.py
"""Supplier-side process for the two-process A2CN HTTP demo."""

from __future__ import annotations

import argparse

import httpx
import uvicorn
from fastapi import HTTPException, Request

from demo_shared import (
    BUYER_DID,
    BUYER_KEY_ID,
    SUPPLIER_DID,
    SUPPLIER_KEY_ID,
    SUPPLIER_PORT,
    SUPPLIER_URL,
    FreshJwtAuth,
    buyer_private_key,
    did_document,
    renewal_terms,
    supplier_agent_info,
    supplier_counteroffer,
    supplier_mandate,
    supplier_private_key,
)
from a2cn import server as server_module

app = server_module.app


def configure_supplier() -> None:
    supplier_key = supplier_private_key()
    buyer_key = buyer_private_key()
    server_module.SERVER_DID = SUPPLIER_DID
    server_module.configure_responder(
        {
            "agent_info": supplier_agent_info(),
            "mandate": supplier_mandate(),
            "deal_types": ["saas_renewal"],
            "max_rounds_by_deal_type": {"saas_renewal": 4},
            "private_key": supplier_key,
        }
    )
    server_module.register_did_document(
        BUYER_DID,
        did_document(BUYER_DID, BUYER_KEY_ID, buyer_key),
    )
    server_module.register_did_document(
        SUPPLIER_DID,
        did_document(SUPPLIER_DID, SUPPLIER_KEY_ID, supplier_key),
    )


configure_supplier()


@app.get("/demo/health")
async def demo_health() -> dict:
    return {"status": "ok", "role": "supplier", "did": SUPPLIER_DID}


@app.post("/demo/counteroffer")
async def demo_counteroffer(request: Request) -> dict:
    body = await request.json()
    session_id = body.get("session_id")
    in_reply_to = body.get("in_reply_to")
    terms = body.get("terms")
    if not session_id or not in_reply_to or not isinstance(terms, dict):
        raise HTTPException(status_code=400, detail="session_id, in_reply_to, and terms are required")

    session = server_module.manager.get_session(session_id)
    if session is None:
        raise HTTPException(status_code=404, detail=f"Unknown session_id {session_id!r}")

    message = supplier_counteroffer(session, terms, in_reply_to)
    supplier_key = supplier_private_key()
    async with httpx.AsyncClient(
        auth=FreshJwtAuth(SUPPLIER_DID, SUPPLIER_DID, supplier_key, SUPPLIER_KEY_ID),
        timeout=10.0,
    ) as http:
        response = await http.post(
            f"{SUPPLIER_URL}/sessions/{session_id}/messages",
            json=message,
            headers={
                "Content-Type": "application/a2cn+json",
                "Idempotency-Key": message["message_id"],
            },
        )
    if response.status_code >= 400:
        raise HTTPException(status_code=response.status_code, detail=response.text)
    return {"message": message, "state": response.json()}


@app.get("/demo/record/{session_id}")
async def demo_record(session_id: str) -> dict:
    session = server_module.manager.get_session(session_id)
    if session is None:
        raise HTTPException(status_code=404, detail=f"Unknown session_id {session_id!r}")
    from a2cn.record import generate_transaction_record

    return generate_transaction_record(session)


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Run the A2CN two-process supplier agent")
    parser.add_argument(
        "--port",
        type=int,
        default=SUPPLIER_PORT,
        help="Port to listen on; must match SUPPLIER_PORT in demo_shared.py",
    )
    args = parser.parse_args()
    uvicorn.run(app, host="127.0.0.1", port=args.port, log_level="warning")

File: a2cn_ts/demos/two_process/supplier_agent.ts
/** Supplier-side process for the two-process A2CN HTTP demo. */

import { createServerContext } from "../../src/a2cn/server.js";
import { generateTransactionRecord } from "../../src/a2cn/record.js";
import type { Dict } from "../../src/a2cn/messages.js";
import {
  BUYER_DID,
  BUYER_KEY_ID,
  SUPPLIER_DID,
  SUPPLIER_KEY_ID,
  SUPPLIER_PORT,
  SUPPLIER_URL,
  authedFetch,
  buyerPrivateKey,
  didDocument,
  freshJwtFactory,
  supplierAgentInfo,
  supplierCounteroffer,
  supplierMandate,
  supplierPrivateKey,
} from "./demo_shared.js";

const ctx = createServerContext();

function configureSupplier(): void {
  const supplierKey = supplierPrivateKey();
  const buyerKey = buyerPrivateKey();
  ctx.SERVER_DID = SUPPLIER_DID;
  ctx.configureResponder({
    agent_info: supplierAgentInfo(),
    mandate: supplierMandate(),
    deal_types: ["saas_renewal"],
    max_rounds_by_deal_type: { saas_renewal: 4 },
    private_key: supplierKey,
  });
  ctx.registerDidDocument(BUYER_DID, didDocument(BUYER_DID, BUYER_KEY_ID, buyerKey));
  ctx.registerDidDocument(SUPPLIER_DID, didDocument(SUPPLIER_DID, SUPPLIER_KEY_ID, supplierKey));
}

configureSupplier();

ctx.app.get("/demo/health", async (_request, reply) => {
  return reply
    .header("content-type", "application/json")
    .send(JSON.stringify({ status: "ok", role: "supplier", did: SUPPLIER_DID }));
});

ctx.app.post("/demo/counteroffer", async (request, reply) => {
  let body: Dict;
  try {
    const raw = request.body;
    body = JSON.parse(Buffer.isBuffer(raw) ? raw.toString("utf-8") : String(raw)) as Dict;
  } catch {
    return reply.code(400).send(
      JSON.stringify({ detail: "session_id, in_reply_to, and terms are required" }),
    );
  }
  const sessionId = body.session_id as string | undefined;
  const inReplyTo = body.in_reply_to as string | undefined;
  const terms = body.terms;
  if (!sessionId || !inReplyTo || terms === null || typeof terms !== "object") {
    return reply.code(400).send(
      JSON.stringify({ detail: "session_id, in_reply_to, and terms are required" }),
    );
  }

  const session = ctx.manager.getSession(sessionId);
  if (session === null) {
    return reply
      .code(404)
      .send(JSON.stringify({ detail: `Unknown session_id ${JSON.stringify(sessionId)}` }));
  }

  const message = supplierCounteroffer(session, terms as Dict, inReplyTo);
  const supplierKey = supplierPrivateKey();
  const http = authedFetch(
    freshJwtFactory(SUPPLIER_DID, SUPPLIER_DID, supplierKey, SUPPLIER_KEY_ID),
  );
  const response = await http(`${SUPPLIER_URL}/sessions/${sessionId}/messages`, {
    method: "POST",
    body: JSON.stringify(message),
    headers: {
      "Content-Type": "application/a2cn+json",
      "Idempotency-Key": message.message_id as string,
    },
  });
  const responseText = await response.text();
  if (response.status >= 400) {
    return reply.code(response.status).send(JSON.stringify({ detail: responseText }));
  }
  return reply.send(JSON.stringify({ message, state: JSON.parse(responseText) }));
});

ctx.app.get("/demo/record/:session_id", async (request, reply) => {
  const { session_id: sessionId } = request.params as { session_id: string };
  const session = ctx.manager.getSession(sessionId);
  if (session === null) {
    return reply
      .code(404)
      .send(JSON.stringify({ detail: `Unknown session_id ${JSON.stringify(sessionId)}` }));
  }
  return reply.send(JSON.stringify(generateTransactionRecord(session)));
});

const portArgIndex = process.argv.indexOf("--port");
const port =
  portArgIndex >= 0 && portArgIndex + 1 < process.argv.length
    ? parseInt(process.argv[portArgIndex + 1], 10)
    : SUPPLIER_PORT;

ctx.app.listen({ host: "127.0.0.1", port }).catch((exc) => {
  console.error(exc);
  process.exit(1);
});
